event.go

```diff
@@ -176,8 +208,33 @@ func (eb *EventBuilder) Build(eventID string, now time.Time, origin ServerName,
 		return
 	}
 
-	result.eventJSON = eventJSON
-	if err = json.Unmarshal(eventJSON, &result.fields); err != nil {
+	result.roomVersion = roomVersion
+
+	switch eventFormat {
+	case EventFormatV1:
+		result.eventJSON = eventJSON
+		var fields eventFormatV1Fields
+		if err = json.Unmarshal(eventJSON, &fields); err != nil {
+			return
+		}
+		fields.fixNilSlices()
+		result.fields = fields
+	case EventFormatV2:
+		if result.eventJSON, err = sjson.DeleteBytes(eventJSON, "event_id"); err != nil {
+			return
+		}
+		var fields eventFormatV2Fields
+		if err = json.Unmarshal(eventJSON, &fields); err != nil {
+			return
+		}
+		fields.EventID, err = result.generateEventID()
+		if err != nil {
+			return
```

Comment:
This function is now too long to really be using named return args...

Reply:
For now I am erring on the side of leaving it as-is since we'd still have to instantiate empty `Event` objects at each error-return.